keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,942 @@
+
+*   This presents a challenge when existing features begin to be reported through the new Node Capabilities mechanism. For such features, we would need a phased capability reliance in the control plane. The presence of the capability `lifecycle.kubernetes.io/node-capabilities: "true"` would indicate if the node itself has opted into the capability framework. Capability-based filtering will only be applied to nodes that have opted into the framework (i.e., those reporting `lifecycle.kubernetes.io/node-capabilities: "true"`). Nodes that have not opted not be filtered out, thus retaining the existing scheduling behavior for them 
+
+
+Scenario 2 : kube-scheduler is looking for certain capability (a new feature added in v1.X) but nodes are running older kubelet versions (v1.X-2) that do not have the capability. Node capabilities feature is GA in v1.X-2 so all nodes are reporting their capabilities.
+
+*   This is a valid use case that Node Capabilities is trying to address. The pod that relies on the feature being enabled should remain pending with a valid status message. 
+*   This prevents pods from failing or running with degraded functionality on incompatible nodes, which can happen today if feature-specific node labels and selectors are not properly applied.
+
+### Future Considerations 
+
+#### Explicit Capability Request 
+
+For the Alpha implementation, a pod's capability requirements are inferred by the shared library based on its specification. While this centralizes the logic, this implicit model still requires a code change in the shared library for each new capability that is introduced. A suggestion from the SIG Scheduling meeting was to explore a mechanism to keep the scheduler logic generic and avoid having case-specific inference logic.
+
+#### Cluster Autoscaler Integration
+
+The cluster-autoscaler makes scale-up decisions based on `NodeGroup` templates, which are abstract representations of a node. Since node capabilities are reported by a running Kubelet, they are not present on these templates. This creates a prediction gap: the autoscaler cannot know what capabilities a new node will have, making it unable to scale up a node pool from zero nodes to satisfy a capability-specific pending pod.
+
+Considered approaches:
+
+1.  Have the autoscaler inspect a running node in the target node pool and assume all new nodes will be identical. This would work only if a running node exists and fails for the "scale-from-zero" conditions.
+2.  This problem is fundamentally the same as what [kubernetes/autoscaler#7799](https://github.com/kubernetes/autoscaler/issues/7799) is tracking to support DRA use cases. The cluster-autoscaler currently does not consider DRA resources while scaling up and the long term solution would likely involve a new API surface to specify and/or modify autoscaler predictions.
```

Comment:
is it the same? I thought that DRA is unique as DRA is not a part of a Node Status so it is harder to add those to the templates

Reply:
It is same in the sense that both NodeCapabilities and DRA are scheduler critical information determined after teh node is created and is not present in the instance template.

Yes, capabilities are present in `node.status` but using it in cluster autoscaler has 2 main drawbacks. 
1. Does not work in scale-from-zero scenario's
2. During upgrade, different nodes within the same group/pool might have different capabilities. This would make it unreliable as the cluster autoscaler samples one node from the group.